test/firrtl/lower-to-rtl-module.mlir

```diff
@@ -210,4 +210,19 @@
     // CHECK: [[OUTF_CAST:%.+]] = firrtl.stdIntCast [[OUTF]]
     // CHECK: rtl.output %inA, [[OUTB]], [[OUTC]], [[OUTD]], [[OUTE_CAST]], [[OUTF_CAST]]
   }
+
+  // CHECK: rtl.module @Queue_1
+  firrtl.module @Queue_1(%clock: !firrtl.clock, %bar: !firrtl.flip<uint<3>>) {
+    // CHECK-NEXT: firrtl.stdIntCast
+    // CHECK-NEXT: firrtl.mem
+    %ram_foo = firrtl.mem "Undefined" {depth = 2 : i64, name = "ram_foo", readLatency = 0 : i32, writeLatency = 1 : i32} : !firrtl.bundle<MPORT: bundle<addr: uint<1>, en: uint<1>, clk: clock, data: uint<3>, mask: uint<1>>, io_MPORT: bundle<addr: uint<1>, en: uint<1>, clk: clock, data: flip<uint<3>>>>
+    // CHECK-NEXT: firrtl.subfield
+    %0 = firrtl.subfield %ram_foo("io_MPORT") : (!firrtl.bundle<MPORT: bundle<addr: uint<1>, en: uint<1>, clk: clock, data: uint<3>, mask: uint<1>>, io_MPORT: bundle<addr: uint<1>, en: uint<1>, clk: clock, data: flip<uint<3>>>>) -> !firrtl.bundle<addr: uint<1>, en: uint<1>, clk: clock, data: flip<uint<3>>>
+    // CHECK-NEXT: firrtl.subfield
+    %1 = firrtl.subfield %0("data") : (!firrtl.bundle<addr: uint<1>, en: uint<1>, clk: clock, data: flip<uint<3>>>) -> !firrtl.flip<uint<3>>
+    // CHECK-NEXT: firrtl.asPassive
+    // CHECK-NEXT: [[OUTG:%.+]] = firrtl.stdIntCast
+    firrtl.connect %bar, %1 : !firrtl.flip<uint<3>>, !firrtl.flip<uint<3>>
```

Comment:
Thanks :-)

Somewhat related to the firrtl.connect verification work, is this supposed to work and do we want to model this in the firrtl dialect?

I look at this sort of flip handling as sort of like lvalue to rvalue conversions in a language like C.  Making them explicit (with an asPassive) operation would make it easier to handle, because we don't have to remember to strip off flip types everywhere.

I implemented this for primitive ops some time ago, and it defined away a lot of bugs.  Before that, primitives could take either normal values or flips of them.

One challenge in the case of firrtl.connect is that it works on aggregates which have interior flips, so literally converting to a passive type on the LHS doesn't work - I think the invariant would be that the LHS and the flipped RHS are compatible with each other?

Reply:
I've been looking at this as well, since I'm looking at the connect verifications.

>  is this supposed to work

This is my question as well. My understanding is roughly: memories have source flow, and subfields have the same flow as the memory, unless the field is flipped. So, `%ram_foo` has source flow, `%0` has source flow, and `%1` has sink flow. `%bar` is an output, and has sink flow. I am not sure the flow of this connect is valid. I haven't studied the part of the spec about memories specifically, just looking at connects and flows. Am I missing something?

More generally, I am thinking about how memories are modele in the FIRRTL dialect. We [discussed](https://github.com/llvm/circt/issues/262) just using multiple results for instances, so maybe we could do something analogous for memories?

Reply:
It's not clear that it is support in the spec (5.1 rule 4), but we are seeing it.  It may be an error in the parser related to memory types.

Reply:
Here is the reduced firrtl testcase:
circuit Queue_1 :

  module Queue_1 :
    input clock : Clock
    output foo : UInt<3>
  
    mem ram_bar
      data-type => UInt<3>
      depth => 2
      read-latency => 0
      write-latency => 1
      reader => foo_MPORT
      writer => MPORT
      read-under-write => undefined
    foo <= ram_bar.foo_MPORT.data

Reply:
Got it. Thanks for the info, and please keep me in the loop.

Reply:
Another instance of this problem for instances:
```
module  {
  firrtl.circuit "Bar" {
    firrtl.module @Foo(%clock: !firrtl.clock, %reset: !firrtl.uint<1>, %io_in: !firrtl.uint<32>) {
    }
    firrtl.module @Bar(%clock: !firrtl.clock, %reset: !firrtl.uint<1>, %io_cpu_resp_bits_raw: !firrtl.flip<uint<32>>) {
    
      %expander = firrtl.instance @Foo {name = "foo"} : !firrtl.bundle<clock: flip<clock>, reset: flip<uint<1>>, io_in: flip<uint<32>>>

      %241 = firrtl.subfield %expander("io_in") : (!firrtl.bundle<clock: flip<clock>, reset: flip<uint<1>>, io_in: flip<uint<32>>>) -> !firrtl.flip<uint<32>>

      firrtl.connect %io_cpu_resp_bits_raw, %241 : !firrtl.flip<uint<32>>, !firrtl.flip<uint<32>>
    }
  }
}
```

Reply:
Talking with @seldridge, this second one is due to chisel allowing reads of input ports on instances as though they were duplex (wires).  Firrtl picks up this behavior, despite clearly being contrary to the spec.
```
import chisel3._
import chisel3.stage.ChiselStage

class Bar extends RawModule {
  val in = IO(Input(Bool()))
}

class Foo extends RawModule {
  val in = IO(Input(Bool()))
  val out = IO(Output(Bool()))
  
  val bar = Module(new Bar)
  
  bar.in := in
  
  out := bar.in
}

println(ChiselStage.emitChirrtl(new Foo))
println(ChiselStage.emitVerilog(new Foo))
```

Reply:
Chisel bug here: https://github.com/chipsalliance/firrtl/issues/2013

Reply:
The first issue is a type inference failure fixed here: https://github.com/llvm/circt/pull/305

Reply:
nice catch!

Reply:
Awesome.

Reply:
The resolution of the Chisel bug is that chisel will emit wires in cases where this appears.  Since this should be resolved upstream soon, I'm closing this.